### PR TITLE
cmake: add compile-warnings-as-errors for c++

### PR DESCRIPTION
**Pull request by jeppenodgaard**
Enabling Kconfig option COMPILER_WARNINGS_AS_ERRORS did not convert warnings to errors in C++ files.

Add a C++ compile option to fix it.

### CODE REVIEW

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by tejlmand — DISMISSED** [review]

**Comment by jeppenodgaard** [comment]
`west build -p -b zybo tests/lib/cpp/cxx -T cpp.main.cpp98` fails with:

```
In file included from /__w/zephyr/zephyr/include/zephyr/arch/arm/arch.h:50,
                 from /__w/zephyr/zephyr/include/zephyr/arch/cpu.h:19,
                 from /__w/zephyr/zephyr/include/zephyr/kernel_includes.h:37,
                 from /__w/zephyr/zephyr/include/zephyr/kernel.h:17,
                 from /__w/zephyr/zephyr/tests/lib/cpp/cxx/src/main.cpp:18:
/__w/zephyr/zephyr/include/zephyr/arch/arm/cortex_a_r/timer.h:40:1: error: this decimal constant is unsigned only in ISO C90 [-Werror]
   40 | DEVICE_MMIO_TOPLEVEL_STATIC(timer_regs, ARM_TIMER_NODE);

```

I am not sure how to fix this warning.

I can change devicetree.h but I am not sure what I am doing:
```
-#define DT_REG_ADDR(node_id) DT_REG_ADDR_BY_IDX(node_id, 0)
+#define DT_REG_ADDR(node_id) UINT32_C(DT_REG_ADDR_BY_IDX(node_id, 0))
...
+#if defined(_LINKER) || defined(_ASMLANGUAGE)
+#define DT_U32_C(_v) (_v)
+#else
+#define DT_U32_C(_v) UINT32_C(_v)
+#endif
```

**Comment by stephanosio** [comment]
> `west build -p -b zybo tests/lib/cpp/cxx -T cpp.main.cpp98` fails with:

FWIW, a related discussion in https://github.com/zephyrproject-rtos/zephyr/issues/53505.

File: tests/lib/cbprintf_package/src/main.c
Comment (cfriedt) [inline tests/lib/cbprintf_package/src/main.c:928]: I'm a bit perplexed as to why this `unsinfed int` cast is necessary. Both `clen` and `slen` are `int`. Perhaps the following might be better?

```suggestion
	zassert_true(clen ==
		     slen + (int)sizeof(test_str1) + 1 /*null*/ - 2 /* arg+ro idx gone*/);
```

Comment (jeppenodgaard) [reply tests/lib/cbprintf_package/src/main.c]: Indeed. Fixed.

**Comment by cfriedt** [comment]
> FWIW, a related discussion in https://github.com/zephyrproject-rtos/zephyr/issues/53505.

This kind of puts undefined behaviour to shame

**Review by tejlmand — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by jeppenodgaard** [comment]
@tejlmand Can I do anything to get this PR in a mergeable state? Or are we stuck until a solution for https://github.com/zephyrproject-rtos/zephyr/issues/53505 is found?

**Comment by henrikbrixandersen** [comment]
> @tejlmand Can I do anything to get this PR in a mergeable state? Or are we stuck until a solution for #53505 is found?

We cannot merge this PR as long as it is not passing the CI checks, if that's what you are asking.

**Comment by kartben** [comment]
rebased to trigger stuck CI

**Comment by cfriedt** [comment]
I read somewhere recently that this warning may be avoided by adding a flexible array member at the end of the structure. E.g.

```cpp
struct foo {
  /* otherwise empty*/

  /* empty struct warning avoidance*/
#ifdef __cplusplus
  uint32_t _[];
#endif
};
```

However, I just tried via godbolt and received an error with `-Wall -Wextra -Werror` flags.

```shell
# Compilation provided by Compiler Explorer at https://godbolt.org/
<Compilation failed>
# Compiler exited with result code 1
Standard error:
<source>:11:12: error: flexible array member 'foo::_' in an otherwise empty 'struct foo'
   11 |   uint32_t _[];
      |            ^
```

**Comment by jeppenodgaard** [comment]
Closing since there's a [duplicate PR](https://github.com/zephyrproject-rtos/zephyr/pull/78419) that looks promising.

**Pull request closed without merging**
